src/kernel/filesystem/fat32.zig

```diff
@@ -0,0 +1,1756 @@
+        ///     IN dir: *const vfs.DirNode   - The parent directory.
+        ///     IN name: []const u8          - The name of the file/folder to open.
+        ///     IN flags: vfs.OpenFlags      - The open flags for if creating a file/folder/symlink.
+        ///     IN open_args: vfs.OpenArgs   - The open arguments if creating a symlink.
+        ///
+        /// Return: *vfs.Node
+        ///     The VFS Node for the opened/created file/folder.
+        ///
+        /// Error: Allocator.Error || ReadError || SeekError || vfs.Error
+        ///     Allocator.Error           - Not enough memory for allocating memory
+        ///     ReadError                 - Errors when reading the stream.
+        ///     SeekError                 - Errors when seeking the stream.
+        ///     vfs.Error.NoSuchFileOrDir - Error if creating a symlink and no target is provided.
+        ///
+        fn createFileOrDirOrSymlink(fs: *const vfs.FileSystem, dir: *const vfs.DirNode, name: []const u8, flags: vfs.OpenFlags, open_args: vfs.OpenArgs) (Allocator.Error || ReadError || SeekError || vfs.Error)!*vfs.Node {
+            const self = @fieldParentPtr(Fat32Self, "instance", fs.instance);
+
+            // TODO: Future PR
+            return vfs.Error.NoSuchFileOrDir;
+        }
+
+        ///
+        /// Deinitialise this file system.
+        ///
+        /// Arguments:
+        ///     IN self: *Fat32Self - Self to free.
+        ///
+        pub fn deinit(self: *Fat32Self) void {
+            // Make sure we have closed all files
+            // TODO: Should this deinit close any open files instead?
+            std.debug.assert(self.opened_files.count() == 0);
+            self.opened_files.deinit();
+            self.allocator.destroy(self.root_node.node);
+            self.allocator.destroy(self.fs);
+            self.allocator.destroy(self);
+        }
+
+        ///
+        /// Initialise a FAT32 filesystem.
+        ///
+        /// Arguments:
+        ///     IN allocator: *Allocator - Allocate memory.
+        ///     IN stream: StreamType - Allocate memory.
+        ///
+        /// Return: *Fat32
+        ///     The pointer to a Fat32 filesystem.
+        ///
+        /// Error: Allocator.Error || ReadError || SeekError || Error
+        ///     Allocator.Error - If there is no more memory. Any memory allocated will be freed.
+        ///     ReadError       - If there is an error reading from the stream.
+        ///     SeekError       - If there si an error seeking the stream.
+        ///     Error           - If there is an error when parsing the stream to set up a fAT32
+        ///                       filesystem. See Error for the list of possible errors.
+        ///
+        pub fn init(allocator: *Allocator, stream: StreamType) (Allocator.Error || ReadError || SeekError || Error)!*Fat32Self {
+            log.info("Init\n", .{});
+            defer log.info("Done\n", .{});
```

Comment:
Fat32 initialisation will be happening after boot initialisation too, so I don't think these log messages are appropriate. We don't want them appearing when the user plugs in a memory stick, for example (we'd want device related log messages instead).

Reply:
Shall i change it to debug logs

Reply:
Yeah that's a good idea